File: src/main/java/com/github/unchama/seichiassist/data/menu/MenuHandler.java
package com.github.unchama.seichiassist.data.menu;

import org.bukkit.entity.EntityType;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryType;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.InventoryHolder;

/**
 * {@link InventoryView} を管理するクラス.
 * Singleton で設計されています.
 *
 * @author karayuu
 */
public final class MenuHandler implements Listener {
    private static MenuHandler singleton = new MenuHandler();

    private MenuHandler() {
    }

    /**
     * {@link MenuHandler} のインスタンスを返します.
     *
     * @return {@link MenuHandler} のインスタンス(singleton)
     */
    public static MenuHandler getInstance() {
        return singleton;
    }

    /**
     * 各 {@link InventoryView#invoke(int, InventoryClickEvent)} を呼び出します.
     *
     * @param event {@link InventoryClickEvent}
     */
    @EventHandler
    public void onInventoryClick(InventoryClickEvent event) {
        if (event.getWhoClicked().getType() != EntityType.PLAYER) {
            return;
        }

        final Inventory clickedInventory = event.getClickedInventory();
        final Inventory openInventory = event.getWhoClicked().getOpenInventory().getTopInventory();
        //メニュー外のクリック排除
        if (clickedInventory == null) {
            return;
        }

        //プレイヤーインベントリ内のクリック排除
        if (openInventory.getHolder() instanceof InventoryView && clickedInventory.getType() == InventoryType.PLAYER) {
            event.setCancelled(true);
            return;
        }

        final InventoryHolder holder = clickedInventory.getHolder();
        if (holder instanceof InventoryView) {
            ((InventoryView) holder).invoke(event.getSlot(), event);
        }
    }
}

File: src/main/java/com/github/unchama/menuinventory/MenuHandler.scala
package com.github.unchama.menuinventory

import com.github.unchama.util.syntax.Nullability.NullabilityExtensionReceiver
import org.bukkit.entity.Player
import org.bukkit.event.inventory.{InventoryClickEvent, InventoryType}
import org.bukkit.event.{EventHandler, Listener}

/**
 * [MenuInventoryView] に由来するインベントリ上のクリックイベントをビューに定義されたアクションに流すようなリスナーオブジェクト.
 *
 * @author karayuu
 */
object MenuHandler extends Listener {
  @EventHandler
  def onInventoryClick(event: InventoryClickEvent) {
    val whoClicked = event.getWhoClicked.refineOrNone[Player].ifNull { return }

    //メニュー外のクリック排除
    val clickedInventory = event.getClickedInventory.ifNull { return }
    val openInventory = event.whoClicked.openInventory.topInventory

    //プレイヤーインベントリ内のクリック排除
    if (openInventory.holder is MenuSession && clickedInventory.getType === InventoryType.PLAYER) {
      event.isCancelled = true
      return
    }

    val holder = clickedInventory.holder

    if (holder is MenuSession) {
      val effect = holder.view.slotLayout.computeAsyncEffectOn(event)

      unsafe {
        runNonBlocking({
          fx {
            !effect {
              effect.runFor(whoClicked)
            }
          }
        }) { if (it is Either.Left) it.a.printStackTrace() }
      }
    }
  }
}
